Fix for #20716: "Our analytics" can be confusing when user does not have access to that collection

calherries opened the pull request:
Fixes https://github.com/metabase/metabase/issues/20716

Basically, "Our analytics" should not be visible from the side menu if the user doesn't have read perms for it.

### Implementation details

Currently a GET request to `/api/collection/root` from a user without root permissions will return a 200 response with details about the root collection. This breaks the FEs expectation that it should only be able to load collections for which it has permissions to read. Any collections requested with `/api/collection/:id` receive a 403 response if the user doesn't have read permissions. Additionally, the FE expects all `/api/collection/root/items` to only contain collections it has permission to read.

So, this change does two things on the BE:
- makes a GET request to `/api/collection/root` return a 403 response if the user doesn't have root permissions
- makes a GET request to `/api/collection/root/items` exclude the root collection if the user doesn't have root permissions

Commit:
Change collection endpoints to not return root collection if user doesn't have Root permission

gusaiani commented:
@calherries, I think I'm missing something about this fix.

Maybe the issue description (in #20716) lacks something, or the testing instructions in this PR could use more details.

To sum it up, what we see in the issue is the old homepage, so it's not clear to me what to look for in our current layout.

calherries commented:
@gusaiani That's a fair point. The original issue is out of date. Nemanja's last comment on the issue contains the relevant up-to-date details. Basically, Our analytics should not be visible from the side menu if the user doesn't have read perms for it.

alxnddr committed:
hide our analytics collection in saved question picker when users do not have access

alxnddr commented:
I updated the saved question picker as well

Commit:
Fix tests

calherries commented:
I'd like some BE feedback on the changes I made to adjust the tests.

See [this commit](https://github.com/metabase/metabase/pull/23887/commits/72c824296fe112ca9b462947e84ae33b7079fb88#diff-a6b09b9ecb5cfb6fb2118db8aa7e7c6f6483f79b4e53860bcc2c5e2697311922L1164), from line 1164 onwards.

I removed all the tests that check a root collection's ancestors are an empty list. It seemed to duplicate the behaviour covered by [this test](https://github.com/metabase/metabase/blob/20717-our-analytics-with-no-read-permission/test/metabase/api/collection_test.clj#L930). The tests were failing because the GET requests to `api/collection/root` are wrapped with `mt/with-non-admin-groups-no-root-collection-perms`, which is called from inside the macroexpansion of [`with-collection-hierarchy`](https://github.com/metabase/metabase/blob/20717-our-analytics-with-no-read-permission/test/metabase/models/collection_test.clj#L136). Now that the GET requests return a 403 error the tests fail. 

My question is - should I have adjusted these tests instead of removing them entirely? What were they testing that might not be covered now?

dpsutton reviewed (approved):
backend looks good to me. Thanks for the PR description. This particular PR isn't super hard to follow along but its a good habit to have and helps load up context even for straightforward fixes like here.

gusaiani commented on frontend/src/metabase/query_builder/components/saved-question-picker/SavedQuestionList.jsx:
```diff
@@ -43,46 +44,50 @@ function SavedQuestionList({
+                    .sort((a, b) =>
+                      a.display_name.localeCompare(b.display_name),
+                    )
+                    .map(t => (
```
I see we didn't really want to refactor anything here, we simply put this block inside a `LoadingWrapper`.

At the same time, a couple ideas that could make this block more palatable.

1. Instead of using `t` as the map's iterated item name, to use `table`. 
2. Extracting what's in this map into a function and reducing nesting depth.

gusaiani reviewed (approved):
Looking good to me.

One minor comment inlined.

paoliniluis commented:
Will this get pushed to 44 or we leave it for 45?

nemanjaglumac commented:
> Will this get pushed to 44 or we leave it for 45?

https://github.com/metabase/metabase/issues/20716#issuecomment-1185555079

calherries commented:
@paoliniluis Yes, it should have been backported. Thanks for catching
